src/orion/client/runner.py

```diff
@@ -0,0 +1,339 @@
+
+        while self.is_running:
+            # Get new trials for our free workers
+            with self.stat.time("sample"):
+                new_trials = self.sample()
+
+            # Scatter the new trials to our free workers
+            with self.stat.time("scatter"):
+                self.scatter(new_trials)
+
+            # Gather the results of the workers that have finished
+            with self.stat.time("gather"):
+                self.gather()
+
+            if self.is_idle:
+                idle_end = time.time()
+                idle_time += idle_end - idle_start
+                idle_start = idle_end
+
+                log.debug(f"Workers have been idle for {idle_time:.2f} s")
+            else:
+                idle_start = time.time()
+                idle_time = 0
+
+            if self.is_idle and idle_time > self.idle_timeout:
+                msg = f"Workers have been idle for {idle_time:.2f} s"
+
+                if self.has_remaining and not self.is_done:
+                    msg = (
+                        f"{msg}; worker has leg room (has_remaining: {self.has_remaining})"
+                        f" and optimization is not done (is_done: {self.is_done})"
+                    )
+
+                raise LazyWorkers(msg)
+
+        return self.trials
+
+    def should_sample(self):
+        """Check if more trials could be generated"""
+
+        if self.is_broken or self.is_done:
+            return 0
+
+        pending = len(self.pending_trials) + self.trials
+        remains = self.max_trials_per_worker - pending
+
+        n_trial = min(self.free_worker, remains)
+        should_sample_more = self.free_worker > 0 and remains > 0
+
+        return int(should_sample_more) * n_trial
+
+    def sample(self):
+        """Sample new trials for all free workers"""
+        n_trial = self.should_sample()
+
+        if n_trial > 0:
+            # the producer does the job of limiting the number of new trials
+            # already no need to worry about it
+            # NB: suggest reserve the trial already
+            new_trials = self._suggest_trials(n_trial)
+            log.debug(f"Sampled {len(new_trials)} new configs")
+            return new_trials
+
+        return []
+
+    def scatter(self, new_trials):
+        """Schedule new trials to be computed"""
+        new_futures = []
+        for trial in new_trials:
+            future = self.client.executor.submit(
+                _optimize, trial, self.fct, self.trial_arg, **self.kwargs
+            )
+            self.pending_trials[future] = trial
+            new_futures.append(future)
+
+        self.futures.extend(new_futures)
+        log.debug("Scheduled new trials")
+
+    def gather(self):
+        """Gather the results from each worker asynchronously"""
+        results = []
+        try:
+            results = self.client.executor.async_get(
+                self.futures, timeout=self.gather_timeout
+            )
+        except KeyboardInterrupt:
```

Comment:
Not tested

Reply:
Done here https://github.com/Epistimio/orion/pull/684/commits/0d390c553d734d53440bff566f85eb124b30152e